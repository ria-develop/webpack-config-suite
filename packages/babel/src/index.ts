export * from './JsxExtensions';
export * from './HotModuleReplacementConfig';
export * from './BabelReactConfig';
export * from './BabelConfig';
