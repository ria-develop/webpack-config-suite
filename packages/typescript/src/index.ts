export * from './TsExtension';
export * from './TsxExtension';
export * from './TypeScriptConfig';
export * from './TypeScriptPlugins';
export * from './TypeScriptReactConfig';
