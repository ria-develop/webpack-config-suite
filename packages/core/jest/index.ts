export {TestCases} from '../../../jest';
import {processConfig, WebpackCliOptions, What} from '../src';
import {Configuration} from 'webpack';

export function testProcessConfig(
  Entity: What,
  config: Configuration,
  env: unknown = undefined,
  argv: WebpackCliOptions = {}
): void {
  const composed = processConfig(Entity, config, env, argv);
  if (composed) {
    composed.forEach((ComposedConfig) => testProcessConfig(ComposedConfig, config, env, argv));
  }
}
