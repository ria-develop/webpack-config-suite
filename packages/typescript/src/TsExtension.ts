import {ExtensionsAwareConfig} from '@webpack-config-suite/core';

export class TsExtension extends ExtensionsAwareConfig {
  protected get extensions(): string[] {
    return ['.ts'];
  }
}
