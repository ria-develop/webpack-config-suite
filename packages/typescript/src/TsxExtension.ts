import {ResolveExtensions} from '@webpack-config-suite/core';

export class TsxExtension extends ResolveExtensions {
  protected get extensions(): string[] {
    return ['.tsx'];
  }
}
