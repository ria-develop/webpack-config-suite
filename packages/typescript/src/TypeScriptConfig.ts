import {TsExtension, TypeScriptPlugins} from './';

import {BaseConfig} from '@webpack-config-suite/core';
import {RuleSetUseItem} from 'webpack';
import {BabelConfig} from '@webpack-config-suite/babel';
import {Options as BabelPresetEnvOptions} from '@babel/preset-env';
import {PluginItem} from '@babel/core';
import {Options as StrictTsLoaderOptions} from 'ts-loader';

export type TsLoaderOptions = Partial<StrictTsLoaderOptions>;

export class TypeScriptConfig extends BabelConfig {
  static BABEL_PRESET_TYPESCRIPT = '@babel/preset-typescript';

  public get composed(): typeof BaseConfig[] {
    return [TsExtension, this.isSingleOrDevMode && TypeScriptPlugins].filter(Boolean);
  }

  //--------------------------------------------------------------------------------------------------------------------
  // Babel Typescript
  //--------------------------------------------------------------------------------------------------------------------

  protected get babelPresetTypescriptOptions(): BabelPresetEnvOptions | null {
    return null;
  }

  protected get babelPresetTypescript(): (string | BabelPresetEnvOptions | null)[] {
    return [TypeScriptConfig.BABEL_PRESET_TYPESCRIPT, this.babelPresetTypescriptOptions].filter(Boolean);
  }

  protected get babelPresets(): PluginItem[] | null {
    return super.babelPresets.concat([this.babelPresetTypescript]);
  }

  protected get test(): RegExp {
    return /\.ts/;
  }

  protected get tsLoaderOptions(): TsLoaderOptions {
    return {transpileOnly: true, happyPackMode: this.useThread};
  }

  protected get tsLoader(): RuleSetUseItem {
    return {loader: 'ts-loader', options: this.tsLoaderOptions};
  }

  protected get use(): RuleSetUseItem[] {
    return super.use.concat([this.tsLoader]);
  }
}
