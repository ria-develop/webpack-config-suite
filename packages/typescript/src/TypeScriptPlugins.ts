import {PluginsAwareConfig} from '@webpack-config-suite/core/';
import {Configuration} from 'webpack';
import {ForkTsCheckerWebpackPlugin} from 'fork-ts-checker-webpack-plugin/lib/ForkTsCheckerWebpackPlugin';
import {ForkTsCheckerWebpackPluginOptions} from 'fork-ts-checker-webpack-plugin/lib/ForkTsCheckerWebpackPluginOptions';

export class TypeScriptPlugins extends PluginsAwareConfig {
  protected get forkTsCheckerOptions(): ForkTsCheckerWebpackPluginOptions {
    return {
      typescript: {
        diagnosticOptions: {
          semantic: true,
          syntactic: true
        }
      },
      eslint: {
        files: './src/**/*.{ts,tsx,js,jsx}',
        enabled: true
      }
    };
  }

  protected get forkTsChecker(): ForkTsCheckerWebpackPlugin {
    return new ForkTsCheckerWebpackPlugin(this.forkTsCheckerOptions);
  }

  get plugins(): Configuration['plugins'] | undefined {
    return (this.isDev && [this.forkTsChecker]) || undefined;
  }
}
