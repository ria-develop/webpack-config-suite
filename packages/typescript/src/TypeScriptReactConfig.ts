import {TypeScriptConfig, TsxExtension} from './';

import {BaseConfig} from '@webpack-config-suite/core';
import {HotModuleReplacementConfig} from '@webpack-config-suite/babel';

export class TypeScriptReactConfig extends TypeScriptConfig {
  public get composed(): typeof BaseConfig[] {
    return super['composed'].concat([TsxExtension, this.isHot && HotModuleReplacementConfig]).filter(Boolean);
  }

  protected get test(): RegExp {
    return /\.tsx?/;
  }

  protected get useReact(): boolean {
    return true;
  }
}
