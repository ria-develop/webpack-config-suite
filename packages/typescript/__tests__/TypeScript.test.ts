import {WebpackCliOptions} from '@webpack-config-suite/core';
import {Configuration} from 'webpack';
import {TypeScriptConfig, TypeScriptReactConfig} from '../src';
import {TestCases} from '../../../jest/TestCases';

describe.each(TestCases)(
  'given mode: "%s", env: "%j", argv: "%j"',
  (mode: Configuration['mode'], env: unknown, argv: WebpackCliOptions): void => {
    it('given TypeScriptConfig should match snapshot', () => {
      const config: Configuration = {mode};
      class MyTypeScriptConfig extends TypeScriptConfig {
        protected get numWorkers(): number {
          return 1;
        }
      }
      new MyTypeScriptConfig(config, env, argv).composed.forEach((Config) => new Config(config, env, argv));
      expect(config).toMatchSnapshot();
    });
    it('given TypeScriptReactConfig should match snapshot', () => {
      const config: Configuration = {mode};
      class MyTypeScriptReactConfig extends TypeScriptReactConfig {
        protected get numWorkers(): number {
          return 1;
        }
      }
      new MyTypeScriptReactConfig(config, env, argv).composed.forEach((Config) => new Config(config, env, argv));
      expect(config).toMatchSnapshot();
    });
  }
);
