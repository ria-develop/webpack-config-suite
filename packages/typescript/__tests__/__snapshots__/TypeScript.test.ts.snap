// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given mode: "development", env: "undefined", argv: "{"liveReload":false,"hot":false}" given TypeScriptConfig should match snapshot 1`] = `
Object {
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.ts/,
        "use": Array [
          Object {
            "loader": "cache-loader",
          },
          Object {
            "loader": "thread-loader",
            "options": Object {
              "poolTimeout": Infinity,
              "workers": 1,
            },
          },
          Object {
            "loader": "babel-loader",
            "options": Object {
              "presets": Array [
                Array [
                  "@babel/preset-env",
                ],
                Array [
                  "@babel/preset-typescript",
                ],
              ],
            },
          },
          Object {
            "loader": "ts-loader",
            "options": Object {
              "happyPackMode": true,
              "transpileOnly": true,
            },
          },
        ],
      },
    ],
  },
  "plugins": Array [
    ForkTsCheckerWebpackPlugin {
      "options": Object {
        "eslint": Object {
          "enabled": true,
          "files": "./src/**/*.{ts,tsx,js,jsx}",
        },
        "typescript": Object {
          "diagnosticOptions": Object {
            "semantic": true,
            "syntactic": true,
          },
        },
      },
    },
  ],
  "resolve": Object {},
}
`;

exports[`given mode: "development", env: "undefined", argv: "{"liveReload":false,"hot":false}" given TypeScriptReactConfig should match snapshot 1`] = `
Object {
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.tsx\\?/,
        "use": Array [
          Object {
            "loader": "cache-loader",
          },
          Object {
            "loader": "thread-loader",
            "options": Object {
              "poolTimeout": Infinity,
              "workers": 1,
            },
          },
          Object {
            "loader": "babel-loader",
            "options": Object {
              "presets": Array [
                Array [
                  "@babel/preset-env",
                ],
                Array [
                  "@babel/preset-react",
                ],
                Array [
                  "@babel/preset-typescript",
                ],
              ],
            },
          },
          Object {
            "loader": "ts-loader",
            "options": Object {
              "happyPackMode": true,
              "transpileOnly": true,
            },
          },
        ],
      },
    ],
  },
  "plugins": Array [
    ForkTsCheckerWebpackPlugin {
      "options": Object {
        "eslint": Object {
          "enabled": true,
          "files": "./src/**/*.{ts,tsx,js,jsx}",
        },
        "typescript": Object {
          "diagnosticOptions": Object {
            "semantic": true,
            "syntactic": true,
          },
        },
      },
    },
  ],
  "resolve": Object {
    "extensions": Array [
      ".tsx",
    ],
  },
}
`;

exports[`given mode: "development", env: "undefined", argv: "{"liveReload":true,"hot":true}" given TypeScriptConfig should match snapshot 1`] = `
Object {
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.ts/,
        "use": Array [
          Object {
            "loader": "cache-loader",
          },
          Object {
            "loader": "thread-loader",
            "options": Object {
              "poolTimeout": Infinity,
              "workers": 1,
            },
          },
          Object {
            "loader": "babel-loader",
            "options": Object {
              "plugins": Array [
                "react-hot-loader/babel",
              ],
              "presets": Array [
                Array [
                  "@babel/preset-env",
                ],
                Array [
                  "@babel/preset-typescript",
                ],
              ],
            },
          },
          Object {
            "loader": "ts-loader",
            "options": Object {
              "happyPackMode": true,
              "transpileOnly": true,
            },
          },
        ],
      },
    ],
  },
  "plugins": Array [
    ForkTsCheckerWebpackPlugin {
      "options": Object {
        "eslint": Object {
          "enabled": true,
          "files": "./src/**/*.{ts,tsx,js,jsx}",
        },
        "typescript": Object {
          "diagnosticOptions": Object {
            "semantic": true,
            "syntactic": true,
          },
        },
      },
    },
  ],
  "resolve": Object {},
}
`;

exports[`given mode: "development", env: "undefined", argv: "{"liveReload":true,"hot":true}" given TypeScriptReactConfig should match snapshot 1`] = `
Object {
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.tsx\\?/,
        "use": Array [
          Object {
            "loader": "cache-loader",
          },
          Object {
            "loader": "thread-loader",
            "options": Object {
              "poolTimeout": Infinity,
              "workers": 1,
            },
          },
          Object {
            "loader": "babel-loader",
            "options": Object {
              "plugins": Array [
                "react-hot-loader/babel",
              ],
              "presets": Array [
                Array [
                  "@babel/preset-env",
                ],
                Array [
                  "@babel/preset-react",
                ],
                Array [
                  "@babel/preset-typescript",
                ],
              ],
            },
          },
          Object {
            "loader": "ts-loader",
            "options": Object {
              "happyPackMode": true,
              "transpileOnly": true,
            },
          },
        ],
      },
    ],
  },
  "plugins": Array [
    ForkTsCheckerWebpackPlugin {
      "options": Object {
        "eslint": Object {
          "enabled": true,
          "files": "./src/**/*.{ts,tsx,js,jsx}",
        },
        "typescript": Object {
          "diagnosticOptions": Object {
            "semantic": true,
            "syntactic": true,
          },
        },
      },
    },
  ],
  "resolve": Object {
    "alias": Object {
      "react-dom": "@hot-loader/react-dom",
    },
    "extensions": Array [
      ".tsx",
    ],
  },
  "target": "web",
}
`;

exports[`given mode: "development", env: "undefined", argv: "{"mode":"development"}" given TypeScriptConfig should match snapshot 1`] = `
Object {
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.ts/,
        "use": Array [
          Object {
            "loader": "cache-loader",
          },
          Object {
            "loader": "thread-loader",
            "options": Object {
              "poolTimeout": Infinity,
              "workers": 1,
            },
          },
          Object {
            "loader": "babel-loader",
            "options": Object {
              "presets": Array [
                Array [
                  "@babel/preset-env",
                ],
                Array [
                  "@babel/preset-typescript",
                ],
              ],
            },
          },
          Object {
            "loader": "ts-loader",
            "options": Object {
              "happyPackMode": true,
              "transpileOnly": true,
            },
          },
        ],
      },
    ],
  },
  "plugins": Array [
    ForkTsCheckerWebpackPlugin {
      "options": Object {
        "eslint": Object {
          "enabled": true,
          "files": "./src/**/*.{ts,tsx,js,jsx}",
        },
        "typescript": Object {
          "diagnosticOptions": Object {
            "semantic": true,
            "syntactic": true,
          },
        },
      },
    },
  ],
  "resolve": Object {},
}
`;

exports[`given mode: "development", env: "undefined", argv: "{"mode":"development"}" given TypeScriptReactConfig should match snapshot 1`] = `
Object {
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.tsx\\?/,
        "use": Array [
          Object {
            "loader": "cache-loader",
          },
          Object {
            "loader": "thread-loader",
            "options": Object {
              "poolTimeout": Infinity,
              "workers": 1,
            },
          },
          Object {
            "loader": "babel-loader",
            "options": Object {
              "presets": Array [
                Array [
                  "@babel/preset-env",
                ],
                Array [
                  "@babel/preset-react",
                ],
                Array [
                  "@babel/preset-typescript",
                ],
              ],
            },
          },
          Object {
            "loader": "ts-loader",
            "options": Object {
              "happyPackMode": true,
              "transpileOnly": true,
            },
          },
        ],
      },
    ],
  },
  "plugins": Array [
    ForkTsCheckerWebpackPlugin {
      "options": Object {
        "eslint": Object {
          "enabled": true,
          "files": "./src/**/*.{ts,tsx,js,jsx}",
        },
        "typescript": Object {
          "diagnosticOptions": Object {
            "semantic": true,
            "syntactic": true,
          },
        },
      },
    },
  ],
  "resolve": Object {
    "extensions": Array [
      ".tsx",
    ],
  },
}
`;

exports[`given mode: "production", env: "undefined", argv: "{"liveReload":true,"hot":true}" given TypeScriptConfig should match snapshot 1`] = `
Object {
  "mode": "production",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.ts/,
        "use": Array [
          Object {
            "loader": "cache-loader",
          },
          Object {
            "loader": "thread-loader",
            "options": Object {
              "poolTimeout": 2000,
              "workers": 1,
            },
          },
          Object {
            "loader": "babel-loader",
            "options": Object {
              "presets": Array [
                Array [
                  "@babel/preset-env",
                ],
                Array [
                  "@babel/preset-typescript",
                ],
              ],
            },
          },
          Object {
            "loader": "ts-loader",
            "options": Object {
              "happyPackMode": true,
              "transpileOnly": true,
            },
          },
        ],
      },
    ],
  },
  "resolve": Object {},
}
`;

exports[`given mode: "production", env: "undefined", argv: "{"liveReload":true,"hot":true}" given TypeScriptReactConfig should match snapshot 1`] = `
Object {
  "mode": "production",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.tsx\\?/,
        "use": Array [
          Object {
            "loader": "cache-loader",
          },
          Object {
            "loader": "thread-loader",
            "options": Object {
              "poolTimeout": 2000,
              "workers": 1,
            },
          },
          Object {
            "loader": "babel-loader",
            "options": Object {
              "presets": Array [
                Array [
                  "@babel/preset-env",
                ],
                Array [
                  "@babel/preset-react",
                ],
                Array [
                  "@babel/preset-typescript",
                ],
              ],
            },
          },
          Object {
            "loader": "ts-loader",
            "options": Object {
              "happyPackMode": true,
              "transpileOnly": true,
            },
          },
        ],
      },
    ],
  },
  "resolve": Object {
    "extensions": Array [
      ".tsx",
    ],
  },
}
`;

exports[`given mode: "production", env: "undefined", argv: "{"mode":"production"}" given TypeScriptConfig should match snapshot 1`] = `
Object {
  "mode": "production",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.ts/,
        "use": Array [
          Object {
            "loader": "cache-loader",
          },
          Object {
            "loader": "thread-loader",
            "options": Object {
              "poolTimeout": 2000,
              "workers": 1,
            },
          },
          Object {
            "loader": "babel-loader",
            "options": Object {
              "presets": Array [
                Array [
                  "@babel/preset-env",
                ],
                Array [
                  "@babel/preset-typescript",
                ],
              ],
            },
          },
          Object {
            "loader": "ts-loader",
            "options": Object {
              "happyPackMode": true,
              "transpileOnly": true,
            },
          },
        ],
      },
    ],
  },
  "plugins": undefined,
  "resolve": Object {},
}
`;

exports[`given mode: "production", env: "undefined", argv: "{"mode":"production"}" given TypeScriptReactConfig should match snapshot 1`] = `
Object {
  "mode": "production",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.tsx\\?/,
        "use": Array [
          Object {
            "loader": "cache-loader",
          },
          Object {
            "loader": "thread-loader",
            "options": Object {
              "poolTimeout": 2000,
              "workers": 1,
            },
          },
          Object {
            "loader": "babel-loader",
            "options": Object {
              "presets": Array [
                Array [
                  "@babel/preset-env",
                ],
                Array [
                  "@babel/preset-react",
                ],
                Array [
                  "@babel/preset-typescript",
                ],
              ],
            },
          },
          Object {
            "loader": "ts-loader",
            "options": Object {
              "happyPackMode": true,
              "transpileOnly": true,
            },
          },
        ],
      },
    ],
  },
  "plugins": undefined,
  "resolve": Object {
    "extensions": Array [
      ".tsx",
    ],
  },
}
`;
